contracts/distribution/dao-rewards-distributor/src/contract.rs

```diff
@@ -0,0 +1,531 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    from_json, to_json_binary, Addr, BankMsg, Binary, Coin, CosmosMsg, Deps, DepsMut, Env,
+    MessageInfo, Response, StdError, StdResult, Uint128, Uint256, WasmMsg,
+};
+use cw2::set_contract_version;
+use cw20::Denom::Cw20;
+use cw20::{Cw20ReceiveMsg, Denom};
+use cw4::MemberChangedHookMsg;
+use dao_hooks::{nft_stake::NftStakeChangedHookMsg, stake::StakeChangedHookMsg};
+use dao_interface::voting::{
+    Query as VotingQueryMsg, TotalPowerAtHeightResponse, VotingPowerAtHeightResponse,
+};
+use std::cmp::min;
+use std::convert::TryInto;
+
+use crate::msg::{
+    ExecuteMsg, InfoResponse, InstantiateMsg, PendingRewardsResponse, QueryMsg, ReceiveMsg,
+};
+use crate::state::{
+    Config, RewardConfig, CONFIG, LAST_UPDATE_BLOCK, PENDING_REWARDS, REWARD_CONFIG,
+    REWARD_PER_TOKEN, USER_REWARD_PER_TOKEN,
+};
+use crate::ContractError;
+use crate::ContractError::{
+    InvalidCw20, InvalidFunds, NoRewardsClaimable, RewardPeriodNotFinished,
+};
+
+const CONTRACT_NAME: &str = env!("CARGO_PKG_NAME");
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // Intialize the contract owner
+    cw_ownable::initialize_owner(deps.storage, deps.api, msg.owner.as_deref())?;
+
+    let reward_denom = match msg.reward_denom {
+        Denom::Native(denom) => Denom::Native(denom),
+        Cw20(addr) => Cw20(deps.api.addr_validate(addr.as_ref())?),
+    };
+
+    // Verify contract provided is a voting module contract
+    let _: TotalPowerAtHeightResponse = deps.querier.query_wasm_smart(
+        &msg.vp_contract,
+        &VotingQueryMsg::TotalPowerAtHeight { height: None },
+    )?;
+
+    // Optional hook caller is allowed to call voting power change hooks.
```

Comment:
can you give an example of how this could be used?

Reply:
For example, cw20 DAOs have separate contracts for staking and voting power. Others, depending on their setup, may also have separate contracts for staking and returning voting power.